### PR TITLE
CTPPS: Improved Diamond data formats

### PR DESCRIPTION
**Pull request by forthommel**
This PR amends the Diamond detector's rechits and local tracks data formats definition (and associated producers) to include the z-coordinate and a time precision container for future calibration developments.
Additionally, an extra check on a leading/trailing edges association is performed on hits collection before building the local tracks, reflecting DAQ-level optimisations performed during 2017 data-taking.
Finally, a helper public method is added to the `CTPPSDiamondLocalTrack` object to retrieve the association between a track and rechits.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by forthommel** [comment]
Dears @slava77, @perrotta,
Please let us know if you need any additional information from our side to ease your review!

**Comment by perrotta** [comment]
@forthommel : I still had no time to go through it in detail.

However, there is something that you could already answer, to speed up further checks:
- A few parameters are modified in the fillDescription method of CTPPSDiamondLocalTrackFitter.cc: 
  - Why did you modify those parameters
  - This is not mentioned in the PR description: please add a few words about it
  - Why there is no effect on the jenkins outputs? Should we run on some different dataset in order to assess the effect of those modified parameters? Can you suggest us how to test it?

File: RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc [hunk 1]
```diff
@@ -11,42 +11,51 @@
...
+    const CTPPSDiamondDetId detid( vec.detId() );
 
-    if ( detid.channel() > 20 ) continue;              // VFAT-like information, to be ignored by CTPPSDiamondRecHitProducer
+    if ( detid.channel() > 20 ) continue; // VFAT-like information, to be ignored
 
+    // retrieve the geometry element associated to this DetID
     const DetGeomDesc* det = geom->getSensor( detid );
+
+    if ( det->parents().empty() )
+      edm::LogWarning("CTPPSDiamondRecHitProducerAlgorithm") << "The geometry element for " << detid << " has no parents. Check the geometry hierarchy!";
```
Comment (perrotta) [inline RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc]: If  ```det->parents().empty()```  the z_pos computed at L35 will make the execution crash.
I would make this a LogError, and then you can assert either on ```!det->parents().empty()``` , or (if you want to continue the process anyhow) define a default/dummy z_pos in that case

File: RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc [hunk 2]
```diff
@@ -11,42 +11,51 @@
...
-      if ( digi->getLeadingEdge()==0 and digi->getTrailingEdge()==0 ) { continue; }
+    for ( const auto& digi : vec ) {
+      if ( digi.getLeadingEdge() == 0 && digi.getTrailingEdge() == 0 ) continue;
 
-      const int t = digi->getLeadingEdge();
-      const int t0 = ( t-t_shift_ ) % 1024;
-      const int time_slice = ( t-t_shift_ ) / 1024;
+      const int t = digi.getLeadingEdge();
+      const int t0 = t % 1024;
+      int time_slice = t / 1024;
```
Comment (perrotta) [inline RecoCTPPS/TotemRPLocal/src/CTPPSDiamondRecHitProducerAlgorithm.cc]: Why not ```const``` any more?

**Comment by nminafra** [comment]
Dear @perrotta,
The sigma parameter is used to parametrize the efficiency at the edges of the detector pads. Practically it means that when two adjacent pads give a signal also the region between them is considered for the track reconstruction, but with a <1 weight. 
capture
A graphic example is shown in the picture: with sigma==0 only 2 hits are participating to the track, while with sigma==0.1 the reco algorithm is considering 4 hits. This gives an idea of how difficult it is to see the effect of sigma!
Putting sigma to 0 means ignoring all signals due to particles passing just outside the detectors pads.
Moreover, if sigma==0, the reco works with integers (zero outside pads, one inside), therefore 1.5 or 1.9 are absolutely equivalent.
I apologize to not have a "quantitative" answer, our approach is to put "safe values" until a complete performance analysis is completed (hopefully as soon as possible).
I might add that we are also working in integrating the time information in the tracks with a future PR that will not impact on the footprint of CTPPSDiamondLocalTrack.

File: RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc [hunk 3]
```diff
@@ -109,12 +103,12 @@ CTPPSDiamondLocalTrackFitter::fillDescriptions( edm::ConfigurationDescriptions&
     ->setComment( "general verbosity of this module" );
 
   edm::ParameterSetDescription trackingAlgoParams;
-  trackingAlgoParams.add<double>( "threshold", 1.5 )
+  trackingAlgoParams.add<double>( "threshold", 1.9 )
```
[the remark this answers is not in the document]
Comment (forthommel) [reply RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc]: Thanks for your fast review, @perrotta ! Please see https://github.com/cms-sw/cmssw/pull/21426#issuecomment-347248769 below for a global comment on these parameters changes.

File: DataFormats/CTPPSReco/interface/CTPPSDiamondLocalTrack.h [hunk 4]
```diff
@@ -19,13 +19,24 @@ class CTPPSDiamondLocalTrack
...
+      num_hits_( 0 ), num_planes_( 0 ), valid_( true ),
+      t_( 0. ), t_sigma_( 0. ), ts_index_( 0 ), mh_( 0 ) {}
     CTPPSDiamondLocalTrack( const math::XYZPoint& pos0, const math::XYZPoint& pos0_sigma, float chisq, float t, float t_sigma, int oot_idx, int mult_hits ) :
       pos0_( pos0 ), pos0_sigma_( pos0_sigma ),
-      chi_squared_( chisq ), valid_( false ),
+      num_hits_( 0 ), num_planes_( 0 ), valid_( false ),
       t_( t ), t_sigma_( t_sigma ), ts_index_( oot_idx ), mh_( mult_hits ) {}
     virtual ~CTPPSDiamondLocalTrack() {}
 
+    inline bool containsHit( const CTPPSDiamondRecHit& recHit, float tolerance = 0.1 ) const {
```
[the remark this answers is not in the document]
Comment (forthommel) [reply DataFormats/CTPPSReco/interface/CTPPSDiamondLocalTrack.h]: This method acts as a helper for the hits ←/→ tracks association retrieval for our timing studies. As a constant public member it does not change the memory footprint of the `CTPPSDiamondLocalTrack` object.

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc [hunk 5]
```diff
@@ -109,12 +103,12 @@ CTPPSDiamondLocalTrackFitter::fillDescriptions( edm::ConfigurationDescriptions&
...
     ->setComment( "minimal number of rechits to be observed before launching the track recognition algorithm" );
   trackingAlgoParams.add<double>( "thresholdFromMaximum", 0.5 );
-  trackingAlgoParams.add<double>( "resolution", 0.01 /* mm */ )
+  trackingAlgoParams.add<double>( "resolution", 0.005 /* mm */ )
```
[the remark this answers is not in the document]
Comment (nminafra) [reply RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc]: "resolution" is the step used for the reco algorithm. We updated the value for a better resolution, but we can also revert this value when including the next set of comments. As previously noted we are preparing a future PR with the correct parameters as soon as we complete the performance characterization.

File: RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc [hunk 6]
```diff
@@ -109,12 +103,12 @@ CTPPSDiamondLocalTrackFitter::fillDescriptions( edm::ConfigurationDescriptions&
...
     ->setComment( "spatial resolution on the horizontal coordinate (in mm)" );
-  trackingAlgoParams.add<double>( "sigma", 0.1 );
+  trackingAlgoParams.add<double>( "sigma", 0 );
```
[the remark this answers is not in the document]
Comment (nminafra) [reply RecoCTPPS/TotemRPLocal/plugins/CTPPSDiamondLocalTrackFitter.cc]: Sorry to not clarify this earlier, please see #21426 (comment) below for a global comment on these parameters changes. Also for this parameter,  we can also revert this value when including the next set of comments, waiting to complete the performance characterization.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
The new members num_hits and num_planes of CTPPSDiamondLocalTrack are always 0: are they just placeholders for future developments? Maybe, there is something missing in the constructor?

**Comment by nminafra** [comment]
@perrotta Indeed: they are placeholders for the next PR when we'll update also the time of the track (t_) and its precision (t_sigma_)

**Comment by perrotta** [comment]
+1
- A couple of dataformats updated:
  - Two integers are added (and one flat is removed) to the DataFormat of CTPPSDiamondLocalTrack.h
  - Three floats are added to the DataFormat of CTPPSDiamondRecHit.h
- By now, the new members num_hits and num_planes of CTPPSDiamondLocalTrack are not yet filled by the producer: a new PR is expected soon with them, and with updates to the time of the track and its precision (https://github.com/cms-sw/cmssw/pull/21426#issuecomment-348920511): given we are now at the beginning of the dev cycle, I personally don't see it as an issue and I think it can even be convenient to factorize the work into the two PRs.
- @forthommel : please open the other PR, or report here about the planes for it

**Comment by davidlange6** [comment]
+1

**Comment by nminafra** [comment]
Thanks @perrotta and @davidlange6  for the review!

Regards

**Pull request merged**